### PR TITLE
 HPCC-21041 Reimplement the code to detect ambiguous selectors

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
The first commit introduces a new way of coping with ambiguous slectors.  It is essentially refactoring the code, but does change 1 example (seep11) because it is slightly less restrictive than the previous code.

Reproducing the potential problem is very hard.  You can use example complexhoist5.ecl, but I'll add details to the jira about how it could have been triggered.

@shamser please review.

**Comment by ghalliday** [comment]
Smoketest failed because it is dependent on another change which I hadn't pushed.  See https://github.com/hpcc-systems/HPCC-Platform/pull/11940.

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -5362,29 +5354,12 @@ inline void addActiveTable(UsedExpressionHashTable & array, IHqlExpression * ds)
...
-    if (numTables == 1)
-    {
-        if (numActiveTables == 0)
-            tables.single->Release();
-    }
-    else if (numTables != 0)
-    {
-        for (unsigned i=numActiveTables; i < numTables; i++)
-            tables.multi[i]->Release();
-        delete [] tables.multi;
```
Comment (shamser) [inline ecl/hql/hqlexpr.cpp]: @ghalliday Possible memory leak: your probably still need the CUsedTables destructor to delete tables.multi?

**Review by shamser — CHANGES_REQUESTED** [review]
@ghalliday Possible memory leak

**Review by shamser — CHANGES_REQUESTED** [review]
@ghalliday

File: ecl/hqlcpp/hqlresource.cpp [hunk 2]
```diff
@@ -419,9 +460,101 @@ class EclChildSplitPointLocator : public EclHoistLocator
...
+            }
+        }
+
+        //If there are too many selectors to maintain a bitset, then set all the bits.
+        if (minHidden && tooManySelectors())
+            childMask = (unsigned)-1;
+    }
+
+    //Check if the expression could be evaluated outside of the current activity
+    void analyseHoistable(IHqlExpression * expr)
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/hqlcpp/hqlresource.cpp]: no, it is correct - analyse if it is able to be hoisted.

**Comment by ghalliday** [comment]
Rebased to force smoketest to run again.

File: ecl/hql/hqlexpr.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp]: no - it is correct.  The records that need to be released are no longer added.

**Comment by ghalliday** [comment]
@shamser replied to comments, and repushed to force smoketest to run.

File: ecl/hql/hqlexpr.cpp [hunk 3]
```diff
@@ -5400,113 +5375,59 @@ bool CUsedTables::usesSelector(IHqlExpression * selector) const
...
-            tables.single = &newTables.item(0);
-            tables.single->Link();
-        }
+        tables.single = &activeTables.item(0);
     }
-    else if (numTables != 0)
+    else if (numActiveTables != 0)
     {
-        IHqlExpression * * multi = new IHqlExpression * [numTables];
+        IHqlExpression * * multi = new IHqlExpression * [numActiveTables];
```
Comment (shamser) [inline ecl/hql/hqlexpr.cpp]: @ghalliday Where is this (new IHqlExpression * [numActiveTables]) released?

File: ecl/hqlcpp/hqlresource.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shamser) [reply ecl/hqlcpp/hqlresource.cpp]: @ghalliday  Any problems in using analyseHoistAble?  Current form looks like it's either analyse Hoist Table or typo?

**Review by shamser — COMMENTED** [review]
@ghalliday Please can you clarify.

File: ecl/hqlcpp/hqlresource.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/hqlcpp/hqlresource.cpp]: I could rename completely, but we don't normally use camel case for the middle of words.

File: ecl/hql/hqlexpr.cpp (hunk 3 above)
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp]: well spotted - I misread your original comment.  I will fix.

**Comment by ghalliday** [comment]
@shamser fixed memory leak and renamed the function

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed. Looks good.

**Pull request merged**